### DISCUSSION TITLE
401 error while using Airflow API

### DISCUSSION THREAD

**Discussion by megha120** [apache/airflow#23755] [General]
### Apache Airflow version

2.1.0

### What happened

Getting the following error while trying out the Airflow API:
{
    "detail": null,
    "status": 401,
    "title": "Unauthorized",
    "type": "https://airflow.apache.org/docs/apache-airflow/2.1.0/stable-rest-api-ref.html#section/Errors/Unauthenticated"
}

### What you think should happen instead

The API should work properly as I am passing the username and password as required for basic authentication.
<img width="540" alt="auth" src="https://user-images.githubusercontent.com/65885522/168012616-ab344035-765c-42ca-a7ee-e608a30099e8.png">

Even with the swagger ui I am getting the same error.
![image](https://user-images.githubusercontent.com/65885522/168012732-55106a4e-dab4-4772-a183-7ddd60cb52be.png)
![image](https://user-images.githubusercontent.com/65885522/168012792-41166ecc-dc40-4d94-96bd-2bf7aafc8333.png)

### How to reproduce

I am using the local installation of Apache Airflow (https://airflow.apache.org/docs/apache-airflow/stable/start/local.html)

### Operating System

Linux

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by ephraimbuddy** [comment]
You need to set the configuration for basic authentication properly. See https://airflow.apache.org/docs/apache-airflow/2.1.0/security/api.html, let me know if it worked for you, if not I will reopen the issue

**Comment by megha120** [comment]
I have set the configuration to airflow.api.auth.backend.basic_auth. You can check in the screenshot below, 
<img width="540" alt="auth" src="https://user-images.githubusercontent.com/65885522/168017997-f3360931-3deb-4ed9-8796-060e2e32f63e.png">

But it is still not working.

**Comment by megha120** [comment]
> You need to set the configuration for basic authentication properly. See https://airflow.apache.org/docs/apache-airflow/2.1.0/security/api.html, let me know if it worked for you, if not I will reopen the issue

@ephraimbuddy It is still not working for me, kindly assist.

**Comment by potiuk** [comment]
I am afraid it does not work this way - if you have some findings and want to discuss it and share with others your errors - feel free to open  discussion and discuss it. Unless you can point out to the exact issue of airlfow - you can just hope on someone who has an experience with setting to help you - but this is not a "bug" of airflow and people here help in their free time/when they have enough knowledge and experience (maybe - there is no guarantee). I think you have some configuration problem - because the authorization works for basically many people (including our tests) - but unless you show logs and set that there is an airflow issue, this is at most a discussion. 

BTW. I think you might have webserver misconfigured. The fact that your CLI works does not mean that the webserver is configured the same way - make sure your configuration is the same everywhere and that your websever properly reads it. Also you should look at the log files to try to understand what happens.

**Comment by ctmuthu** [comment]
I'm facing the same issue. I have the config in place. I'm using 2.2.5 Version

**Reply by potiuk** [reply]
If you provide logs and more information, then it increases the chance that someone will investigate it and fix. If you don't - it doesn't.  Even slighlty.

**Reply by zachik250** [reply]
I have the same issue.
In airflow.cfg I have: auth_backends = airflow.api.auth.backend.basic_auth
run the following: 
curl -X GET --user "zachik:zachik11" http://x.x.x.x:8080/api/v1/pools"
with user I created : airflow users create -e zachik@sqreamtech.com -f zachi -l kochman -r Admin -u zachik -p zachik11

or with the default user I can see in the startup of airflow, and getting : 
{
  "detail": null,
  "status": 401,
  "title": "Unauthorized",
  "type": "https://airflow.apache.org/docs/apache-airflow/2.4.1/stable-rest-api-ref.html#section/Errors/Unauthenticated"
}

Same is from postman.

Where I can see the logs? I can only see what is printed on the screen where I start the server ...
and logs of the dags run

**Reply by ctmuthu** [reply]
Sorry for not updating earlier. I also had auth_backends but when I changed from **auth_backends to auth_backend**. It started authorizing. The plural form was the problem.

**Reply by zachik250** [reply]
Thanks for the replay.
I changed from auth_backends to auth_backend 
auth_backend = airflow.api.auth.backend.basic_auth

restart the server.

Same issue with both users :-(

**Reply by zachik250** [reply]
If I change the url - I can see that the page cannot be found - with mean that when I'm with correct URL I realy get to the server but having auth issue.

Where can I found the server logs (not the dags log?)
What else can be the issue ?

**Reply by ctmuthu** [reply]
401 means your username and password has some issue. In my case, the auth_backend config was the problem. 

I guess your config looks like this 

[api]
auth_backend = airflow.api.auth.backend.basic_auth

Do you have another user to test?

**Reply by zachik250** [reply]
I have the user I created manually , and the admin user that I can see when the server is start up - both not working

**Reply by potiuk** [reply]
@zachik250  - I think yiou simply need to debug it - likely you have some error in configuration - maybe another typo. It's possible that you just step in with the debugger over Airlfow webserver (same as any other python application and try to debug it). That's what I'd do, but I can't - you are the only one with the superpowers of debugging your configuration.

Airflow is open-source and that gives you the extra option that you can just check what's going on - I do it all the time with other 3rd-party libraries.

**Reply by liurida** [reply]
> I have the user I created manually , and the admin user that I can see when the server is start up - both not working

after creating a test account with Admin/User/Op roles, I could access the api with the new account.

**Comment by SarveshGuptaPythonist** [comment]
This way worked for me

I added a .env file and added in docker compose file.
```
env_file:
    - .env
```
Like this in all the services like redis and webservers
```
AIRFLOW__API__AUTH_BACKENDS="airflow.api.auth.backend.basic_auth" 
```
- added this in the .env file and working good, rebuild and run.
